base/linalg/dense.jl

```diff
@@ -310,15 +310,23 @@ kron(a::AbstractMatrix, b::AbstractVector)=kron(a,reshape(b,length(b),1))
 kron(a::AbstractVector, b::AbstractMatrix)=kron(reshape(a,length(a),1),b)
 
 # Matrix power
+function ^{T}(A::AbstractMatrix{T}, p::Integer)
+    # non-inversible matrix to negative power is undefined.
+    if (p < 0) && (det(A) == 0)
+        throw(ArgumentError("Negative powers of singular matrix are not defined"))
+    end
+    return p < 0 ? Base.power_by_squaring(inv(A),-Integer(p)) : Base.power_by_squaring(A,Integer(p))
+end
+
 function ^{T}(A::AbstractMatrix{T}, p::Real)
     # non-inversible matrix to negative power is undefined.
     if (p < 0) && (det(A) == 0)
         throw(ArgumentError("Negative powers of singular matrix are not defined"))
     end
 
     # For integer powers, use repeated squaring
-    if isinteger(p)
-        return p < 0 ? Base.power_by_squaring(inv(A),-Integer(p)) : Base.power_by_squaring(A,Integer(p))
+    if isinteger(p) && !(typeof(p) <: Int64)
+        return float.(A^Integer(p))
```

Comment:
This is not right.  First, `Matrix{Integer}^Integer` should be an integer matrix.   Second, for `Matrix{Int}^Float64` we need to convert to floating point and *then* exponentiate in order to avoid overflow problems.  Third, calling `float.(X)` will make an extra copy even if the matrix is already floating point.

We want something like:
```jl
convert(Matrix{promote_op(^, eltype(A), typeof(p)}, A)^Integer(p)
```
although that isn't quite right either: if `A` is not a `Matrix`, but is some other `AbstractMatrix` type, then we probably want the same type.  Maybe
```jl
T = Base.promote_op(^, eltype(A), typeof(p))
return T == eltype(A) ? A^Integer(p) : copy!(similar(A, T), A)^Integer(p)
```